sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -125,7 +133,9 @@ case class GroupExpression(children: Seq[Expression]) extends Expression {
  * so that the proper type conversions can be performed in the analyzer.
  */
 trait ExpectsInputTypes {
+  self: Expression =>
 
   def expectedChildTypes: Seq[DataType]
 
+  override def validInputTypes: Boolean = children.map(_.dataType) == expectedChildTypes
```

Comment:
Actually I don't like the handling of `ExpectsInputTypes`: we do type casting anyway, then if something is wrong, error will be reported for `Cast`, not the root cause -- wrong input types for a expression.